Source/Core/Core/IOS/USB/Emulated/LogitechMic.cpp

```diff
@@ -0,0 +1,732 @@
+    0x10,             /* bBitResolution */
+    0x05,             /* bSamFreqType */
+    0x40, 0x1f, 0x00, /* tSamFreq 1 */
+    0x11, 0x2b, 0x00, /* tSamFreq 2 */
+    0x22, 0x56, 0x00, /* tSamFreq 3 */
+    0x44, 0xac, 0x00, /* tSamFreq 4 */
+    0x80, 0xbb, 0x00, /* tSamFreq 5 */
+
+    /* Endpoint - Standard Descriptor */
+    0x09,       /* bLength */
+    0x05,       /* bDescriptorType */
+    0x84,       /* bEndpointAddress */
+    0x0d,       /* bmAttributes */
+    0x60, 0x00, /* wMaxPacketSize */
+    0x01,       /* bInterval */
+    0x00,       /* bRefresh */
+    0x00,       /* bSynchAddress */
+
+    /* Endpoint - Audio Streaming */
+    0x07,      /* bLength */
+    0x25,      /* bDescriptorType */
+    0x01,      /* bDescriptor */
+    0x01,      /* bmAttributes */
+    0x02,      /* bLockDelayUnits */
+    0x01, 0x00 /* wLockDelay */
+};
+
+int LogitechMic::SubmitTransfer(std::unique_ptr<CtrlMessage> cmd)
+{
+  // Reference: https://www.usb.org/sites/default/files/audio10.pdf
+  DEBUG_LOG_FMT(IOS_USB,
+                "[{:04x}:{:04x} {}:{}] Control: bRequestType={:02x} bRequest={:02x} wValue={:04x}"
+                " wIndex={:04x} wLength={:04x}",
+                m_vid, m_pid, m_index, m_active_interface, cmd->request_type, cmd->request,
+                cmd->value, cmd->index, cmd->length);
+  switch (cmd->request_type << 8 | cmd->request)
+  {
+  case USBHDR(DIR_DEVICE2HOST, TYPE_STANDARD, REC_DEVICE, REQUEST_GET_DESCRIPTOR):
+  {
+    // It seems every game always pokes this at first twice; one with a length of 9 which gets the
+    // config, and then another with the length provided by the config.
+    DEBUG_LOG_FMT(IOS_USB, "[{:04x}:{:04x} {}:{}] REQUEST_GET_DESCRIPTOR index={:04x} value={:04x}",
+                  m_vid, m_pid, m_index, m_active_interface, cmd->index, cmd->value);
+    cmd->FillBuffer(FULL_DESCRIPTOR.data(), std::min<size_t>(cmd->length, FULL_DESCRIPTOR.size()));
+    cmd->GetEmulationKernel().EnqueueIPCReply(cmd->ios_request, IPC_SUCCESS);
+    break;
+  }
+  case USBHDR(DIR_HOST2DEVICE, TYPE_STANDARD, REC_INTERFACE, REQUEST_SET_INTERFACE):
+  {
+    DEBUG_LOG_FMT(IOS_USB, "[{:04x}:{:04x} {}:{}] REQUEST_SET_INTERFACE index={:04x} value={:04x}",
+                  m_vid, m_pid, m_index, m_active_interface, cmd->index, cmd->value);
+    if (static_cast<u8>(cmd->index) != m_active_interface)
+    {
+      const int ret = ChangeInterface(static_cast<u8>(cmd->index));
+      if (ret < 0)
+      {
+        ERROR_LOG_FMT(IOS_USB, "[{:04x}:{:04x} {}:{}] Failed to change interface to {}", m_vid,
+                      m_pid, m_index, m_active_interface, cmd->index);
+        return ret;
+      }
+    }
+    const int ret = SetAltSetting(static_cast<u8>(cmd->value));
+    if (ret == 0)
+      cmd->GetEmulationKernel().EnqueueIPCReply(cmd->ios_request, cmd->length);
+    return ret;
+  }
+  case USBHDR(DIR_DEVICE2HOST, TYPE_CLASS, REC_INTERFACE,
+              static_cast<u8>(RequestCode::RequestGetCur)):
```

Comment:
Sorry, I indirectly made you ugly those case labels.

If you define a little function like:
```cpp
static constexpr u16 LogitechUSBHDR(u8 dir, u8 type, u8 recipient, RequestCode request)
{
  return USBHDR(dir, type, recipient, Common::ToUnderlying(request));
}
```

Then you can de-uglify those `case` labels like:

```suggestion
  case LogitechUSBHDR(DIR_DEVICE2HOST, TYPE_CLASS, REC_INTERFACE, RequestCode::RequestGetCur):
```

Reply:
No worries, I'm new to the project so I'm 100% anticipating having to iterate to get things right, but we'll eventually get to that ideal implementation so more feedback's always better. 👍